sample-viewmodel/src/main/java/com/example/molecule/viewmodel/MoleculeViewModel.kt

```diff
@@ -20,31 +20,45 @@ import androidx.compose.ui.platform.AndroidUiDispatcher
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import app.cash.molecule.RecompositionMode.ContextClock
-import app.cash.molecule.launchMolecule
+import app.cash.molecule.moleculeFlow
+import kotlin.time.Duration.Companion.seconds
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.SharingStarted
 import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.WhileSubscribed
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.stateIn
 
-abstract class MoleculeViewModel<Event, Model> : ViewModel() {
+abstract class MoleculeViewModel<Event, Model>(
+  initialState: Model,
+  started: SharingStarted = SharingStarted.WhileSubscribed(5.seconds),
+  presenter: @Composable (seed: Model, events: Flow<Event>) -> Model,
+) : ViewModel() {
   private val scope = CoroutineScope(viewModelScope.coroutineContext + AndroidUiDispatcher.Main)
 
   // Events have a capacity large enough to handle simultaneous UI events, but
   // small enough to surface issues if they get backed up for some reason.
   private val events = MutableSharedFlow<Event>(extraBufferCapacity = 20)
 
+  private var seed: Model = initialState
+
   val models: StateFlow<Model> by lazy(LazyThreadSafetyMode.NONE) {
-    scope.launchMolecule(mode = ContextClock) {
-      models(events)
-    }
+    moleculeFlow(mode = ContextClock) {
+      presenter(seed, events)
+    }.onEach {
+      seed = it
+    }.stateIn(
+      scope = scope,
+      started = started,
+      initialValue = seed,
```

Comment:
This whole pattern really feels like it should be encapsulated in something.

I could see it being one (or more) of the following:
- A valve. These existed for RxJava and basically let you control the upstream subscription based on another stream. In this case, the lifecycle would be filtered and mapped to a true/false stream based on whether the activity was resumed and that would control the valve on the upstream collection.
- Replaying share. Another RxJava thing, where an observable instance would cache the latest seen value, replay it to new subscribers, and unsubscribe upstream when the subscriber count dropped to 0.
- A custom frame clock (or frame clock wrapper) which only ticks when the lifecycle is resumed. The flow collector would always be present, but instead of using active collection to control whether recomposition occurs we use the frame clock.

I legitimately cannot figure out what to do with this PR. The use case is certainly valid. But the amount of code required to accomplish it as-is seems like it would overwhelm someone trying to learn how to use Molecule with AndroidX View Model rather than guide them.

Reply:
(This is not really a reflection of your PR, more of the current library design, the design of View Model, and the general sadness of lifecycle.)

Reply:
> This is not really a reflection of your PR , more of the current library design, the design of View Model, and the general sadness of lifecycle.

Yes I understand and I agree. It really shouldn't be this way, but it is the current reality unfortunately.

> The use case is certainly valid. But the amount of code required to accomplish it as-is seems like it would overwhelm someone trying to learn how to use Molecule with AndroidX View Model rather than guide them.

So my entire reasoning behind starting this PR came down to the steps we took to adopt molecule in [the app](https://github.com/HedvigInsurance/android/blob/4cb58b21285b548e5541e9bc038d3875f855a3d7/app/molecule/molecule-android/src/main/kotlin/com/hedvig/android/molecule/android/MoleculeViewModel.kt#L19) I work for:

We were looking into adopting Molecule for our codebase for some time now.
1. A long time ago there was no viewmodel sample, and I looked briefly into it and I though "eh, it's probably gonna be hell to do in a ViewModel" so I dropped the idea.
2. Fast forward some time, the sample is there, and I look into it again. It all works using this tiny bit of glue code so I'm thinking let's go ahead and use it in this way exactly as it seems to work well.
3. I realize that the state flow is always hot, keeping potentially expensive flows hot on the backstack for all entries in the backstack which starts becoming a big deal. I pause this idea briefly
4. Then I tried to look for a way to make these two worlds play well together, hence this PR was made, where Tunji was very kind to help me through understanding how to work around ViewModel to make it happen.

Basically then, I fear that people who use AAC ViewModel will go up to step #2 here, and then stop there without realizing there's more to be done. Then eventually when they figure out what happens, they'll blame Molecule instead of ViewModel since that's when this change would have been introduced to them.
So I thought if the sample played well with the lifecycle in the first place, all the people who would come here can see the easy and straightforward sample from the `sample` module here. And if they want a proper integration with ViewModel, they can look into the more involved `sample-viewmodel` module in this repo.

If we can do something to make this PR even better and make the code more understandable, perhaps from the points you suggest I'd love for that to happen of course. Other than that I personally still feel like it's a net positive if the sample which a lot of people are probably going to follow blindly is actually playing well with AAC ViewModel.

Reply:
@JakeWharton, if we consider `LaunchedEffect` this gets even gnarlier. Those are likely the pieces that you'd actually want to shut down — recomposition is designed not to have so many side effects.

On reflection, I think the happy places to land are likely to be:

* Include `rememberSaveable` support, then kill everything with a `stateIn` type solution, without any `seed`-type mechanism
* Just keep the molecule running, killing it when the ViewModel dies
* Include support for a `rememberRetained`, like Circuit has

The last two are most appealing to me. But even `rememberRetained` has weird consequences — IIRC one of the points of retention in ViewModel is to allow those long-running processes to keep going. The only approach we have that would achieve that would be to just keep it running.

And is keeping it running really so bad? How many would actually be kept running at one time in the VM framework? Is it just the VMs currently on screen during config change, or have I fallen off the boat here?

Reply:
It's all the VMs in screens in the backstack, so potentially many of them.

Reply:
It's worth noting that a `rememberSaveable` solution would still be supsceptible to the overwrite on restart issue as collecting from a flow in the presenter requires an initial value.

Whatever was saved will be immediately overwritten by the value returned by the molecule on restart.

Reply:
> It's worth noting that a rememberSaveable solution would still be supsceptible to the overwrite on restart issue if the value to be restored is not parcelable.

Yep. `launchMolecule` should be capable of supporting all of `stateIn`'s parameters, which would fix that issue in a nice way.